pkg/aws/download.go

```diff
@@ -0,0 +1,121 @@
+/*
+Copyright (C) 2021-2023, Kubefirst
+
+This program is licensed under MIT.
+See the LICENSE file for more details.
+*/
+package aws
+
+import (
+	"fmt"
+	"os"
+	"sync"
+
+	"github.com/kubefirst/kubefirst-api/pkg/providerConfigs"
+	"github.com/kubefirst/runtime/pkg"
+	"github.com/kubefirst/runtime/pkg/downloadManager"
+	"github.com/rs/zerolog/log"
+)
+
+func DownloadTools(awsConfig *providerConfigs.ProviderConfig, kubectlClientVersion string, terraformClientVersion string) error {
+
+	log.Info().Msg("starting downloads...")
+
+	// create folder if it doesn't exist
+	err := pkg.CreateDirIfNotExist(awsConfig.ToolsDir)
+	if err != nil {
+		return err
+	}
+
+	errorChannel := make(chan error)
+	wgDone := make(chan bool)
+	// create a waiting group (translating: create a queue of functions, and only pass the wg.Wait() function down
+	// bellow after all the wg.Add(3) functions are done (wg.Done)
+	var wg sync.WaitGroup
+	wg.Add(2)
+
+	go func() {
+
+		kubectlDownloadURL := fmt.Sprintf(
+			"https://dl.k8s.io/release/%s/bin/%s/%s/kubectl",
+			kubectlClientVersion,
+			pkg.LocalhostOS,
+			pkg.LocalhostARCH,
+		)
+		log.Info().Msgf("Downloading kubectl from: %s", kubectlDownloadURL)
+		err = downloadManager.DownloadFile(awsConfig.KubectlClient, kubectlDownloadURL)
+		if err != nil {
+			errorChannel <- err
+			return
+		}
+
+		err = os.Chmod(awsConfig.KubectlClient, 0755)
+		if err != nil {
+			errorChannel <- err
+			return
+		}
+
+		kubectlStdOut, kubectlStdErr, err := pkg.ExecShellReturnStrings(awsConfig.KubectlClient, "version", "--client=true", "-oyaml")
+		log.Info().Msgf("-> kubectl version:\n\t%s\n\t%s\n", kubectlStdOut, kubectlStdErr)
+		if err != nil {
+			errorChannel <- fmt.Errorf("failed to call kubectlVersionCmd.Run(): %v", err)
+			return
+		}
+		wg.Done()
+		log.Info().Msg("Kubectl download finished")
+	}()
+
+	go func() {
+
+		terraformDownloadURL := fmt.Sprintf(
+			"https://releases.hashicorp.com/terraform/%s/terraform_%s_%s_%s.zip",
+			terraformClientVersion,
+			terraformClientVersion,
+			pkg.LocalhostOS,
+			pkg.LocalhostARCH,
+		)
+		log.Info().Msgf("Downloading terraform from %s", terraformDownloadURL)
+		terraformDownloadZipPath := fmt.Sprintf("%s/terraform.zip", awsConfig.ToolsDir)
+		err = downloadManager.DownloadFile(terraformDownloadZipPath, terraformDownloadURL)
+		if err != nil {
+			errorChannel <- fmt.Errorf("error downloading terraform file, %v", err)
+			return
+		}
+
+		downloadManager.Unzip(terraformDownloadZipPath, awsConfig.ToolsDir)
+
+		err = os.Chmod(awsConfig.ToolsDir, 0777)
```

Comment:
not sure if 777 is a good idea 🤔 , 0755 (rwxr-xr-x) the actual user can read/write/execute, but other user just read

Reply:
Good feedback.  thats a good todo